fvm/transaction_processor_test.go

```diff
@@ -0,0 +1,257 @@
+		))
+
+		procFrozen := Transaction(&flow.TransactionBody{Script: deployContract, Authorizers: []flow.Address{frozenAddress}, Payer: frozenAddress}, 0)
+		procNotFrozen := Transaction(&flow.TransactionBody{Script: deployContract, Authorizers: []flow.Address{notFrozenAddress}, Payer: notFrozenAddress}, 0)
+		deployContext := NewContext(zerolog.Nop(), WithServiceAccount(false), WithRestrictedDeployment(false), WithCadenceLogging(false))
+		deployTxInvocator := NewTransactionInvocator(zerolog.Nop())
+		deployRt := runtime.NewInterpreterRuntime()
+
+		deployVm := New(deployRt)
+
+		err := deployTxInvocator.Process(deployVm, deployContext, procFrozen, st)
+		require.NoError(t, err)
+		err = deployTxInvocator.Process(deployVm, deployContext, procNotFrozen, st)
+		require.NoError(t, err)
+
+		// both contracts should load now
+		context := NewContext(log, WithCadenceLogging(true))
+
+		code := func(a flow.Address) []byte {
+			return []byte(fmt.Sprintf(`
+				import Whatever from 0x%s
+	
+				transaction {
+					execute {
+						Whatever.say()
+					}
+				}
+			`, a.String()))
+		}
+
+		// code from not frozen loads fine
+		proc := Transaction(&flow.TransactionBody{Script: code(frozenAddress)}, 0)
+
+		err = txInvocator.Process(vm, context, proc, st)
+		require.NoError(t, err)
+		require.Len(t, proc.Logs, 1)
+		require.Contains(t, proc.Logs[0], "Düsseldorf")
+
+		proc = Transaction(&flow.TransactionBody{Script: code(notFrozenAddress)}, 0)
+
+		err = txInvocator.Process(vm, context, proc, st)
+		require.NoError(t, err)
+		require.Len(t, proc.Logs, 1)
+		require.Contains(t, proc.Logs[0], "Düsseldorf")
+
+		// freeze account
+		err = accounts.SetAccountFrozen(frozenAddress, true)
+		require.NoError(t, err)
+
+		// make sure freeze status is correct
+		frozen, err := accounts.GetAccountFrozen(frozenAddress)
+		require.NoError(t, err)
+		require.True(t, frozen)
+
+		frozen, err = accounts.GetAccountFrozen(notFrozenAddress)
+		require.NoError(t, err)
+		require.False(t, frozen)
+
+		// loading code from frozen account triggers error
+		proc = Transaction(&flow.TransactionBody{Script: code(frozenAddress)}, 0)
+
+		err = txInvocator.Process(vm, context, proc, st)
+		require.Error(t, err)
+
+		// find frozen account specific error
+		require.IsType(t, runtime.Error{}, err)
+		err = err.(runtime.Error).Err
+
+		require.IsType(t, &runtime.ParsingCheckingError{}, err)
+		err = err.(*runtime.ParsingCheckingError).Err
+
+		require.IsType(t, &sema.CheckerError{}, err)
+		checkerErr := err.(*sema.CheckerError)
+
+		checkerErrors := checkerErr.ChildErrors()
+
+		require.Len(t, checkerErrors, 2)
+		require.IsType(t, &sema.ImportedProgramError{}, checkerErrors[0])
+
+		importedCheckerErrors := checkerErrors[0].(*sema.ImportedProgramError).CheckerError.Errors
+		require.Len(t, importedCheckerErrors, 1)
+
+		require.IsType(t, &AccountFrozenError{}, importedCheckerErrors[0])
+		require.Equal(t, frozenAddress, importedCheckerErrors[0].(*AccountFrozenError).Address)
+	})
+}
```

Comment:
@turbolent do you know if there is any better way to assert this? (I'm only interested in the very last exception being somewhere there).

Reply:
You can shorten the first part: `runtime.Error` and `runtime.ParsingCheckingError` both implement `Wrapper`, i.e. have the `Unwrap` function. This allows you to just use `errors.As`:

```go
var checkerErr *sema.CheckerError
if !error.As(t, err, &checkerErr) {
	t.Failf("error is not a checker error: %v", err)
	return
}
```

BTW: Testify 1.7.0 added a helper for this: https://github.com/stretchr/testify/commit/95a9d909e98735cd8211dfc5cbbb6b8b0b665915. I use this helper right now in Cadence:
```go
func RequireErrorAs(t *testing.T, err error, target interface{}) {
	require.True(
		t,
		errors.As(err, target),
		"error chain must contain a %T",
		target,
	)
}
```

For the second part, maybe don't assert for `checkerErrors[0]` specifically, and instead loop over all errors and look for an imported program error (and fail if you can't find it).

And again for the last part, maybe loop over `importedCheckerErrors` instead of asserting for `importedCheckerErrors[0]`.